sdks/python/apache_beam/io/iobase.py

```diff
@@ -942,12 +966,20 @@ def expand(self, pcoll):
                            | core.WindowInto(window.GlobalWindows())
                            | core.GroupByKey()
                            | 'Extract' >> core.FlatMap(lambda x: x[1]))
+    # PreFinalize should run before FinalizeWrite, and the two should not be
+    # fused.
+    pre_finalize_coll = do_once | 'PreFinalize' >> core.FlatMap(
+        _pre_finalize,
+        self.sink,
+        AsSingleton(init_result_coll),
+        AsIter(write_result_coll))
```

Comment:
<!--new_thread; commit:24c81c1018b47c5316f09f3b1eef8725b5492848; resolved:0-->
Probably safer to add a reshuffle here in case DFE decides to re-execute the side input in case of a failure.

Reply:
ack